python/lsst/ip/diffim/subtractImages.py

```diff
@@ -617,6 +605,232 @@ def _sourceSelector(self, sources):
+        """Preconvolve the science image with its own PSF,
+        convolve the template image with a PSF-matching kernel and subtract
+        from the preconvolved science image.
+
+        Parameters
+        ----------
+        template : `lsst.afw.image.ExposureF`
+            The template image, which has previously been warped to
+            the science image. The template bbox will be padded by a few pixels
+            compared to the science bbox.
+        science : `lsst.afw.image.ExposureF`
+            The science exposure.
+        sources : `lsst.afw.table.SourceCatalog`
+            Identified sources on the science exposure. This catalog is used to
+            select sources in order to perform the AL PSF matching on stamp
+            images around them.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            ``scoreExposure`` : `lsst.afw.image.ExposureF`
+                The image difference.
+            ``matchedTemplate`` : `lsst.afw.image.ExposureF`
+                The warped and PSF-matched template exposure.
+        """
+
+        selectSources = self.prepareInputs(template, science, sources,
+                                           finalizedPsfApCorrCatalog=finalizedPsfApCorrCatalog)
+
+        subtractResults = self.runConvolveTemplate(template, science, selectSources)
+
+        return subtractResults
+
+    def runConvolveTemplate(self, template, science, selectSources):
+        """Convolve the template image with a PSF-matching kernel and subtract
+        from the science image.
+
+        Parameters
+        ----------
+        template : `lsst.afw.image.ExposureF`
+            Template exposure, warped to match the science exposure.
+        science : `lsst.afw.image.ExposureF`
+            Science exposure to subtract from the template.
+        selectSources : `lsst.afw.table.SourceCatalog`
+            Identified sources on the science exposure. This catalog is used to
+            select sources in order to perform the AL PSF matching on stamp
+            images around them.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+
+            ``difference`` : `lsst.afw.image.ExposureF`
+                Result of subtracting template and science.
+                Attached PSF is that of the science image.
+            ``matchedTemplate`` : `lsst.afw.image.ExposureF`
+                Warped and PSF-matched template exposure.
+                Attached PSF is that of the science image.
+            ``backgroundModel`` : `lsst.afw.math.Function2D`
+                Background model that was fit while solving for the PSF-matching kernel
+            ``psfMatchingKernel`` : `lsst.afw.math.Kernel`
+                Kernel used to PSF-match the template to the science image.
+        """
+
+        bbox = lsst.geom.Box2I(science.getBBox().getBegin(), science.getBBox().getDimensions())
```

Comment:
Why can't you just say `bbox = lsst.geom.Box2I(science.getBBox()` here, since all you're doing is making a copy? Or maybe even just `bbox = copy(science.getBBox())`?